pkg/controller/engine.go

```diff
@@ -187,15 +187,33 @@ func (e *Engine) Start(name string, o controller.Options, w ...Watch) error {
 // NamedController is a controller that's not yet started. It gives access to
 // the underlying cache, which may be used e.g. to add indexes.
 type NamedController interface {
+	// Start the named controller. Start does not block.
 	Start(ctx context.Context) error
+	// GetCache returns the cache used by the named controller. If the
+	// controller hasn't been started, neither has the cache.
+	//
+	// Warning: Be careful when calling methods with a context before the
+	// controller has been started. The pass context will be used to e.g. start
+	// informers, and the life-cycle will differ from that of the controller. In
+	// particular, use IndexField instead to add indexes to the cache before the
+	// controller is started.
 	GetCache() cache.Cache
+	IndexField(obj client.Object, field string, extractValue client.IndexerFunc)
 }
 
 type namedController struct {
 	name string
 	e    *Engine
 	ca   cache.Cache
 	ctrl controller.Controller
+
+	indexes []index
+}
+
+type index struct {
```

Comment:
Would it make sense to export this type and have Start take a variadic array of it? I'm not sure how this is called, so I'm happy if it's more ergonomic to use the method approach instead.

e.g.:

```go
c.Start(ctx,
    Index{Object: o, Field: "spec.hi", ExtractValue: ifn},
    Index{Object: o, Field: "spec.hello", ExtractValue: ifn2})
```